TMIT November (#100)
tikv/website: content/blog/monthly-november-2019.md

```rust
let guard = pprof::ProfilerGuard::new(100).unwrap();
// ...
if let Ok(report) = guard.report().build() {
    let file = File::create("flamegraph.svg").unwrap();
    report.flamegraph(file).unwrap();
};
```

{{< figure
    src="/img/blog/monthly-2019/flamegraph.png"
    caption="Generated flamegraph"
    number="" >}}

## Reading materials

* In [Case study: TiKV in JD Cloud](https://www.cncf.io/blog/2019/11/26/case-study-tikv-in-jd-cloud/), the CNCF talked to JD Cloud about how they use more than 10 TiKV clusters. The largest hosts 20 billion rows, and some clusters see workloads of 40k (50% read/ 50% write) operations per second with latencies under 10 milliseconds.
* PingCAP CTO and TiKV founding contributor [@c4pt0r] wrote about how TiKV scales horizonally effectively in [Building a Large-scale Distributed Storage System Based on Raft](https://pingcap.com/blog/building-a-large-scale-distributed-storage-system-based-on-raft/).

## Notable PRs

November was a busy month, here's some highlights:

> * [@nrc] submitted a great refactoring in [#5857](https://github.com/tikv/tikv/pull/5857), [#5935](https://github.com/tikv/tikv/pull/5935) and [#5964](https://github.com/tikv/tikv/pull/5964), helping us make TiKV more clean and understanable.

Reviewer: ```suggestion
* [@nrc] submitted a great refactoring PR in [#5857](https://github.com/tikv/tikv/pull/5857), [#5935](https://github.com/tikv/tikv/pull/5935) and [#5964](https://github.com/tikv/tikv/pull/5964), helping us make TiKV more clean and understanable.
```

* [@little-wallace] submitted a PR to optimize the conflict check in `prewrite` in [#5846](https://github.com/tikv/tikv/pull/5846).

> * [@little-wallace] also submitted an latches schheduler optimization in [#6094]https://github.com/tikv/tikv/pull/6094).

Reviewer: ```suggestion
* [@little-wallace] also submitted a latches scheduler optimization in [#6094]https://github.com/tikv/tikv/pull/6094).
```

* Numerous vectorization PRs were submitted, enabling even faster coprocessor execution.
* [@wangwangwar] introduced in-place byte encoding in the codec, helping reduce the number of allocations needed in [#6061](https://github.com/tikv/tikv/pull/6061)
* [@brson] moved more code into engine traits as part of our ongoing engine abstraction in [#5790](https://github.com/tikv/tikv/pull/5790) and [#5901](https://github.com/tikv/tikv/pull/5901).
* [@breeswish] taught TiKV about `WriteRef` which helps avoid allocations in `PointGetter` and `ForwardScanner`.
* [@sticnarf] opened a PR to unify our read pools into a single multi-level thread pool in [#5828](https://github.com/tikv/tikv/pull/5828). Work is still early stage, but this PR is very exciting!
* [@hoverbear] implemented a suggestion from [@siddontang] to reduce our binary sizes while maintaining our backtraces and debugging info in [#5820](https://github.com/tikv/tikv/pull/5820).
* [@hunterlxt] taught TiKV to use GRPC's memory quota features in [#5818](https://github.com/tikv/tikv/pull/5818).
* [@glorv] and [@hoverbear] are working on packaging TiKV.

## Notable issues

> > TLA+ was firstly used by PingCAP in 2017, to verify a variant of Percolator transaction model is feasible. Different from original Percolator, TiDB optimizes its efficiency by supporting primary key and secondary keys prewriting concurrently in prewrite phase. It was a problematic idea at first glance, but with TLA+ the case was clearly identified and targeted changes were applied to enable this optimization. Another example is TLA+ verified the correctness of multi-raft region merge algorithm, which gives confidence before implementing and summarized a more clear documentation. These work was made by a 4-member group and proved a worthy investing afterwards.

Reviewer: Use the below version... The author asked me to edit it days ago but I forgot to reply. 

TLA+ was firstly used by PingCAP in 2017 to verify the feasibility of a variant of the Percolator transaction model. Different from the original Percolator, TiDB’s model has improved efficiency greatly by supporting primary key and secondary key prewriting concurrently in the prewrite phase. The implementation was problematic at first. Thanks to TLA+, the root cause was clearly identified and targeted changes were applied to enable this optimization. Another application of TLA+ at PingCAP is the correctness verification of the multi-raft region merge algorithm, which provided the necessary confidence before we implemented the feature. There were only 4 members invested in the TLA+ related work, but the rewards are far beyond that.

Engineers at PingCAP believe that TLA+ is necessary when it comes to verifying the correctness of distributed systems and specifying the right system behaviors to avoid future patching. They agree that TLA+ is better at expressing the subtle details which natural languages are not good at.

PingCAP has maintained a repo in Github that records all specifications they wrote and keeps aligning TLA+ specifications with the newly implemented database optimizations. Up to now, there're still active PRs and discussions on verifying new features and bugfixes. Speeches and posts on technical reflections and experience of TLA+ application have been shared with the infra community by PingCAP.

>
> PingCAP has maintained a repo in Github recording all specifications they wrote, and keeps aligning TLA+ specifications with newly implemented database optimization. There're still active PRs and disccussions to verify new feature and fix bug from high-level.
>

> > Engineers here agreed that TLA+ is necessary when proving the correctness of distributed systems, specifing the behavior to avoid patching in future, expressing the subtle details which natural languages not good at. Speech and posts were shared with infra community by PingCAP to discuss on technical reflection and experience of TLA+ application.

Reviewer: Use the below version... The author asked me to edit it days ago but I forgot to reply.

> TLA+ was firstly used by PingCAP in 2017 to verify the feasibility of a variant of the Percolator transaction model. Different from the original Percolator, TiDB’s model has improved efficiency greatly by supporting primary key and secondary key prewriting concurrently in the prewrite phase. The implementation was problematic at first. Thanks to TLA+, the root cause was clearly identified and targeted changes were applied to enable this optimization. Another application of TLA+ at PingCAP is the correctness verification of the multi-raft region merge algorithm, which provided the necessary confidence before we implemented the feature. There were only 4 members invested in the TLA+ related work, but the rewards are far beyond that.

> Engineers at PingCAP believe that TLA+ is necessary when it comes to verifying the correctness of distributed systems and specifying the right system behaviors to avoid future patching. They agree that TLA+ is better at expressing the subtle details which natural languages are not good at.

> PingCAP has maintained a repo in Github that records all specifications they wrote and keeps aligning TLA+ specifications with the newly implemented database optimizations. Up to now, there're still active PRs and discussions on verifying new features and bugfixes. Speeches and posts on technical reflections and experience of TLA+ application have been shared with the infra community by PingCAP.

---
title: This Month in TiKV - November 2019
date: 2019-12-04
author: Ana Hobden
---

As winter descends on the northern hemisphere, we encourage you to keep warm by compiling TiKV! In November we had new minor releases, a new crate release, and some community updates!

Let's take a look!

## Releases

This month our team made 2 minor TiKV releases!

You can review the changelogs here:

* [2.1.18](https://github.com/tikv/tikv/releases/tag/v2.1.18)
* [3.0.6](https://github.com/tikv/tikv/releases/tag/v3.0.6)

Upgrading? Things to note:

* In 3.0.6, [#5697](https://github.com/tikv/tikv/pull/5697) introduced the ability to generate flamegraphs of TiKV at runtime! This means the new [pprof](https://crates.io/crates/pprof) crate has reached stable!
* In 3.0.6, [#5769](https://github.com/tikv/tikv/pull/5769) introduced the ability to modify GC IO limits dynamically with `tikv-ctl`.
* In 3.0.6, [#5440](https://github.com/tikv/tikv/pull/5440) there are new metrics around commit logs duration.

The other changes were minor bugfixes.

## New Community Groups

> fter several discussions, our community has voted to adopt policies to support the founding of Special Interest Groups (SIGS) and Working Groups (WGs). Long-term SIGs are intended to focus on certain components of TiKV, as well as WGs to focus on short-term, cross-component projects or goals.

Reviewer: ```suggestion
After several discussions, our community has voted to adopt policies to support the founding of Special Interest Groups (SIGS) and Working Groups (WGs). Long-term SIGs are intended to focus on certain components of TiKV, as well as WGs to focus on short-term, cross-component projects or goals.
```

Author: Oh no

> After several discussions, our community has voted to adopt policies to support the founding of Special Interest Groups (SIGS) and Working Groups (WGs). Long-term SIGs are intended to focus on certain components of TiKV, as well as WGs to focus on short-term, cross-component projects or goals.

Reviewer 2: ```suggestion
After several discussions, our community has voted to adopt policies to support the founding of Special Interest Groups (SIGS) and Working Groups (WGs). Long-term SIGs are intended to focus on certain components of TiKV; WGs will focus on short-term, cross-component projects or goals.
```
